src/coreclr/src/jit/valuenum.cpp

```diff
@@ -4332,6 +4337,51 @@ bool ValueNumStore::IsVNUnsignedCompareCheckedBound(ValueNum vn, UnsignedCompare
     return false;
 }
 
+//------------------------------------------------------------------------
+// IsVNUnsignedCompareConstCheckBound: Checks if the specified vn represents an expression
+//    such as "(uint)cnst <= (uint)len" that implies constant indexes less then cnst are valid
+//
+// Arguments:
+//    vn - Value number to query
+//    info - Pointer to an UnsignedCompareCheckedBoundInfo object to return information about
+//           the expression. Not populated if the vn expression isn't suitable (e.g. i <= len).
+//           This enables optCreateJTrueBoundAssertion to immediatly create an OAK_NO_THROW
+//           assertion instead of the OAK_EQUAL/NOT_EQUAL assertions created by signed compares
+//           (IsVNCompareCheckedBound, IsVNCompareCheckedBoundArith) that require further processing.
+bool ValueNumStore::IsVNUnsignedCompareConstCheckBound(ValueNum vn, UnsignedCompareCheckedBoundInfo* info)
+{
+    VNFuncApp funcApp;
+
+    if (GetVNFunc(vn, &funcApp))
+    {
+        if ((funcApp.m_func == VNF_LE_UN) || (funcApp.m_func == VNF_GT_UN))
+        {
+            // We have (uint)constant <= (uint)len or it's negation (uint)constant > (uint)len
+            if (IsVNPositiveInt32Constant(funcApp.m_args[0]) && IsVNCheckedBound(funcApp.m_args[1]))
+            {
+                info->vnIdx   = funcApp.m_args[0];
+                info->cmpOper = funcApp.m_func;
+                info->vnBound = funcApp.m_args[1];
+                return true;
+            }
+        }
+        else if ((funcApp.m_func == VNF_GE_UN) || (funcApp.m_func == VNF_LT_UN))
+        {
+            // We have (uint)len >= (uint)constant or it's negation (uint)len < (uint)constant
+            if (IsVNCheckedBound(funcApp.m_args[0]) && IsVNPositiveInt32Constant(funcApp.m_args[1]))
```

Comment:
Why does this if stmt have a different ordering that the one above?

`if (IsVNPositiveInt32Constant(funcApp.m_args[0]) && IsVNCheckedBound(funcApp.m_args[1]))`
vs.
`if (IsVNCheckedBound(funcApp.m_args[0]) && IsVNPositiveInt32Constant(funcApp.m_args[1]))`

They both should be ordered so that the cheaper function is called first. which I believe is `IsVNPositiveInt32Constant`